prometheus-to-sd/config/dynamic_source.go

```diff
@@ -71,33 +87,81 @@ func createOptionsForPodSelection(nodeName string, sources map[string]url.URL) v
-	return v1.ListOptions{
+	labelSelector := fmt.Sprintf("%s in (%s)", componentNameLabel, nameList)
+	listOptions := v1.ListOptions{
 		FieldSelector: fmt.Sprintf("spec.nodeName=%s", nodeName),
 		LabelSelector: labelSelector,
 	}
+	glog.Infof("Filters for dynamic components: %+v", listOptions)
+	return listOptions
 }
 
-func getConfigsFromPods(pods []core.Pod, sources map[string]url.URL) []SourceConfig {
-	var sourceConfigs []SourceConfig
-	for _, pod := range pods {
-		componentName := pod.Labels[nameLabel]
-		source, _ := sources[componentName]
-		podConfig := PodConfig{
-			PodId:       pod.Name,
-			NamespaceId: pod.Namespace,
-		}
-		sourceConfig, err := mapToSourceConfig(componentName, source, pod.Status.PodIP, podConfig)
-		if err != nil {
-			glog.Warning("could not create source config for pod %s: %v", pod.Name, err)
-		}
-		sourceConfigs = append(sourceConfigs, *sourceConfig)
+func getConfigFromPod(pod core.Pod, sources map[string]url.URL) (*SourceConfig, error) {
+	if pod.Status.PodIP == "" {
+		return nil, errors.New("pod ip is missing")
+	}
+	componentName := pod.Labels[componentNameLabel]
+	source, _ := sources[componentName]
+	podConfig := PodConfig{
+		PodId:       pod.Name,
+		NamespaceId: pod.Namespace,
 	}
-	return sourceConfigs
+	return mapToSourceConfig(componentName, source, pod.Status.PodIP, podConfig)
 }
 
 func mapToSourceConfig(componentName string, url url.URL, ip string, podConfig PodConfig) (*SourceConfig, error) {
 	port := url.Port()
 	whitelisted := url.Query().Get("whitelisted")
 	return newSourceConfig(componentName, ip, port, url.Path, whitelisted, podConfig)
 }
+
+func getSourceConfigChangeChannel(kubeApi clientset.Interface, sources map[string]url.URL, podOptions v1.ListOptions) (<-chan SourceConfigChange, error) {
+	watchInterface, err := kubeApi.CoreV1().Pods(podNamespace).Watch(podOptions)
+	if err != nil {
+		return nil, err
+	}
+	sourceConfigChannel := make(chan SourceConfigChange)
+	eventChannel := watchInterface.ResultChan()
+	go func() {
+		for event := range eventChannel {
+			sourceConfigChange, err := mapWatchEventToSourceConfigChange(event, sources)
+			if err != nil {
+				glog.Warningf("dropping event because %v", err)
+			} else {
+				sourceConfigChannel <- *sourceConfigChange
+			}
+		}
+	}()
+	return sourceConfigChannel, nil
+}
+
+func mapWatchEventToSourceConfigChange(event watch.Event, sources map[string]url.URL) (*SourceConfigChange, error) {
+	if event.Type == watch.Error {
+		return nil, fmt.Errorf("watch error with event object %+v", event.Object)
+	}
+	pod, ok := event.Object.(*core.Pod)
+	if !ok {
+		return nil, fmt.Errorf("not a pod object %+v", event.Object)
+	}
+	sourceConfig, err := getConfigFromPod(*pod, sources)
+	if err != nil {
+		return nil, fmt.Errorf("could not map pod %s to source config: %v", pod.Name, err)
+	}
+	sourceConfigChange := SourceConfigChange{
+		ChangeType:   getChangeType(event.Type),
+		SourceConfig: *sourceConfig,
```

Comment:
What happens when you get Delete event with pod without IP?

Reply:
I was thinking about that, but not really seeing as possible, is it possible?